torch/distributed/_tensor/random.py

```diff
@@ -99,210 +59,242 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
-    mesh = spec.mesh
-    dim_map = spec.dim_map
-
-    # Compute shard coordinate:
-    # The coordinate on each tensor dim is a tuple (idx, range)
-    # If a DTensor is partitioned on its dim i into n shards, and the current rank
-    # holds the j-th, then its shard coordinate will be (idx=j, range=n) on dim i
-    coordinate = mesh.get_coordinate()
-    assert coordinate is not None
-    shard_coord = [coordinate[mesh_dim] if mesh_dim >= 0 else 0 for mesh_dim in dim_map]
-    shard_size = [mesh.size(mesh_dim) if mesh_dim >= 0 else 1 for mesh_dim in dim_map]
-
-    # compute shard linear index
-    shard_linear_idx = _calc_shard_linear_idx(shard_coord, shard_size)
-
-    # compute starting offset using the first shard's size
-    local_size_on_rank_0 = list(dtensor_shape)
-    for idx, placement in enumerate(spec.placements):
-        if isinstance(placement, Shard):
-            mesh_dim_size = mesh.size(idx)
-            shard_dim = placement.dim
-            local_size_on_rank_0[shard_dim] = placement._local_shard_size_on_dim(
-                dtensor_shape[shard_dim],
-                mesh_dim_size,
-                0,
-                return_offset=False,
-            )[0]
-
-    local_size = math.prod(local_size_on_rank_0)
-
-    # get current RNG offset
-    current_offset = _get_rng_offset(mesh)
-
-    # pytorch: offset must be multiple of 4
-    # source: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp
-    offset_incr = (shard_linear_idx * local_size + 3) // 4 * 4
-    _set_rng_offset(current_offset + offset_incr, mesh)
-
-
-def set_post_op_offset(spec: DTensorSpec, old_offset: int) -> None:
-    """Sets the RNG to a synchronized state after running the local random op. Every
-    rank should set its RNG offset to `old_offset + DTensor.numel()` where old_offset is
-    the offset before calling `set_pre_op_offset` i.e. the offset before running DTensor
-    random ops.
+    @property
+    def rng_states(self) -> Dict[str, Tensor]:
+        return self._states
 
-    Args:
-        spec (:class:`DTensorSpec`): the spec of the DTensor object on which
-            we post-process the offset for running random ops.
+    @property
+    def distribute_region_enabled(self) -> bool:
+        return self._use_distribute_region
 
-    Returns:
-        None
-    """
-    dtensor_shape = spec.shape
-    mesh = spec.mesh
-    numel = math.prod(dtensor_shape)
-    # pytorch: offset must be multiple of 4
-    # source: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp
-    numel = (numel + 3) // 4 * 4
-    _set_rng_offset(old_offset + numel, mesh)
+    @distribute_region_enabled.setter
+    def distribute_region_enabled(self, value) -> None:
+        self._use_distribute_region = value
 
+    def rng_state_is_sync(self, name) -> bool:
+        return name in self.rng_states
 
-def _get_rng_offset(device_mesh: DeviceMesh) -> int:
-    """Returns the random number generator state offset for the calling rank.
+    def reset(self) -> None:
```

Comment:
it looks like this `reset` does not get used anywhere, shall we delete it?

Reply:
True. I'll remove it.